src/phase-2/scWETHv2.sol

```diff
@@ -0,0 +1,533 @@
+            repayWithdrawParams: repayWithdrawParams,
+            supplyBorrowParams: new SupplyBorrowParam[](0),
+            wstEthToWethSwapData: wstEthToWethSwapData,
+            wethToWstEthSwapData: ""
+        });
+
+        // take flashloan
+        _flashLoan(totalFlashLoanAmount, params);
+
+        emit DisInvested(repayWithdrawParams);
+    }
+
+    /// @notice reallocate funds between protocols (without any slippage)
+    // @param wstEthSwapAmount: amount of wstEth to swap to weth
+    function reallocate(
+        RepayWithdrawParam[] calldata from,
+        SupplyBorrowParam[] calldata to,
+        bytes calldata wstEthToWethSwapData,
+        bytes calldata wethToWstEthSwapData
+    ) external {
+        onlyKeeper();
+        uint256 totalFlashLoanAmount;
+        for (uint256 i; i < from.length; i++) {
+            totalFlashLoanAmount += from[i].repayAmount;
+        }
+
+        RebalanceParams memory params = RebalanceParams({
+            repayWithdrawParams: from,
+            supplyBorrowParams: to,
+            wstEthToWethSwapData: wstEthToWethSwapData,
+            wethToWstEthSwapData: wethToWstEthSwapData
+        });
+
+        // take flashloan
+        _flashLoan(totalFlashLoanAmount, params);
+
+        emit Reallocated(from, to);
+    }
+
+    //////////////////// VIEW METHODS //////////////////////////
+
+    function isSupported(uint256 _adapterId) public view returns (bool) {
+        return protocolAdapters.contains(_adapterId);
+    }
+
+    function getAdapter(uint256 _adapterId) external view returns (address adapter) {
+        (, adapter) = protocolAdapters.tryGet(_adapterId);
+    }
+
+    /// @notice returns the total assets (WETH) held by the strategy
+    function totalAssets() public view override returns (uint256 assets) {
+        // value of the supplied collateral in eth terms using chainlink oracle
+        assets = totalCollateral();
+
+        // subtract the debt
+        assets -= totalDebt();
+
+        // add float
+        assets += asset.balanceOf(address(this));
+    }
+
+    /// @notice returns the total assets supplied as collateral (in WETH terms)
+    function totalCollateral() public view returns (uint256 collateral) {
+        uint256 n = protocolAdapters.length();
+        address adapter;
+        for (uint256 i; i < n; i++) {
+            (, adapter) = protocolAdapters.at(i);
+            collateral += IAdapter(adapter).getCollateral(address(this));
+        }
+
+        collateral = oracleLib.wstEthToEth(collateral);
+    }
+
+    /// @notice returns the total ETH borrowed
+    function totalDebt() public view returns (uint256 debt) {
+        uint256 n = protocolAdapters.length();
+        address adapter;
+        for (uint256 i; i < n; i++) {
+            (, adapter) = protocolAdapters.at(i);
+            debt += IAdapter(adapter).getDebt(address(this));
+        }
+    }
+
+    //////////////////// EXTERNAL METHODS //////////////////////////
```

Comment:
don't really need this separator since a lot of functions above are also external